src/utils/jwt.ts

```diff
@@ -0,0 +1,240 @@
+        callback(new Error(NO_MATCHING_KID_ERROR_MESSAGE));
+      } else {
+        callback(null, publicKeys[kid]);
+      }
+    })
+      .catch(error => {
+        callback(error);
+      });
+  }
+}
+
+export class EmulatorSignatureVerifier implements SignatureVerifier {
+  public verify(token: string): Promise<void> {
+    const error = new JwtError(JwtErrorCode.INVALID_TOKEN,
+      'The provided token has invalid signature.');
+
+    // Signature checks skipped for emulator; no need to fetch public keys.
+    return isSignatureValid(token, '')
+      .then(isValid => {
+        return isValid ? Promise.resolve() : Promise.reject(error);
+      });
+  }
+}
+
+function isSignatureValid(token: string, secretOrPublicKey: jwt.Secret | jwt.GetPublicKeyOrSecret,
+  options?: jwt.VerifyOptions): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    jwt.verify(token, secretOrPublicKey, options,
+      (error: jwt.VerifyErrors | null) => {
+        if (!error) {
+          return resolve(true);
+        }
+        if (error.name === 'TokenExpiredError') {
+          return reject(new JwtError(JwtErrorCode.TOKEN_EXPIRED,
+            'The provided token has expired. Get a fresh token from your ' +
+            'client app and try again.'));
+        } else if (error.name === 'JsonWebTokenError') {
+          const prefix = 'error in secret or public key callback: ';
+          if (error.message && error.message.includes(prefix)) {
+            const message = error.message.split(prefix).pop() || 'Error fetching public keys.';
+            const code = (message === NO_MATCHING_KID_ERROR_MESSAGE) ? JwtErrorCode.NO_MATCHING_KID :
+              JwtErrorCode.INVALID_ARGUMENT;
+            return reject(new JwtError(code, message));
+          }
+          return resolve(false);
+        }
+        return reject(new JwtError(JwtErrorCode.INVALID_TOKEN, error.message));
+      });
+  });
+}
+
+/**
+ * Decodes general purpose Firebase JWTs.
+ */
+export function decodeJwt(jwtToken: string): Promise<DecodedToken> {
+  if (!validator.isString(jwtToken)) {
+    return Promise.reject(new JwtError(JwtErrorCode.INVALID_ARGUMENT,
+      'The provided token must be a string.'));
+  }
+
+  const fullDecodedToken: any = jwt.decode(jwtToken, {
+    complete: true,
+  });
+
+  if (!fullDecodedToken) {
+    return Promise.reject(new JwtError(JwtErrorCode.INVALID_ARGUMENT,
+      'Decoding token failed.'));
+  }
+
+  const header = fullDecodedToken?.header;
+  const payload = fullDecodedToken?.payload;
+  return Promise.resolve({ header, payload });
+}
+
+/**
+ * Jwt error code structure.
+ *
+ * @param code The error code.
+ * @param message The error message.
+ * @constructor
+ */
+export class JwtError extends Error {
+  constructor(readonly code: JwtErrorCode, readonly message: string) {
+    super(message);
+    (this as any).__proto__ = JwtError.prototype;
+  }
+}
+
+/**
+ * JWT error codes.
+ */
+export enum JwtErrorCode {
```

Comment:
Let's define a specific `INVALID_SIGNATURE` error code for signature verification errors.